src/http/data_loader.rs

```diff
@@ -42,57 +49,95 @@ impl<C: HttpClient + Send + Sync + 'static + Clone> Loader<DataLoaderRequest> fo
     keys: &[DataLoaderRequest],
   ) -> async_graphql::Result<HashMap<DataLoaderRequest, Self::Value>, Self::Error> {
     if let Some(group_by) = self.batched.clone() {
-      let mut keys = keys.to_vec();
-      keys.sort_by(|a, b| a.to_request().url().cmp(b.to_request().url()));
+      // store the preused length of our keys.
+      let original_key_length = keys.len();
 
       let mut request = keys[0].to_request();
+
       let first_url = request.url_mut();
 
-      for key in &keys[1..] {
-        let request = key.to_request();
-        let url = request.url();
+      let urls = keys[1..].iter().map(|key: &DataLoaderRequest| {
+        let key_request_result = key.to_request();
+        tokio::task::spawn(async move {
+          let url = key_request_result.url().to_owned();
+          url
```

Comment:
I get what your concern is here that overhead of context switching applied, Iam I right here ? 

I was thinking about from the perspective of huge volume of operations needed to be done, since theoretically that could be potentially in multiples of 1000s so keeping all of that on a single thread won't do much of the benefits though!

also as per my understanding here, green threads should eliminate the problems associated with normal OS threads since they have `&mut Context<'_>` and executor internally polling them to completion(Iam not abs. sure here if scheduler actually does something diff. than that? ) therefore spawning multiple green threads on Tokio should scale well and handle the instructions present within async block better way, but maybe what I can do here better is divide the series of multiple computation blocks under a defined number of actual CPU core available on the platform(the value of which can be obtained from config file) and split it accordingly something like this 

![image](https://github.com/tailcallhq/tailcall/assets/45896902/bf769d96-07ca-457c-97d0-59375116e719)

```rust
let config_cores = 6; // just for example, let's say we have 6 logical cores.
let mut allHandles = vec![];
...
let mut keys: &[DataLoaderRequest] = ... ; 
let green_thread_task_factor = (keys.len() / config_cores) ;  // get the factor for amount of work we can divide here .. 
...
let mut start = 0;
for factor in 0..green_thread_task_factor { 
   let end_point = (factor * config_core); 
   let remain_key_len = keys[start..].len();
   
   //compute the remainder block here ! 
   if remain_key_len < end_point && (factor == gren_thread_task_factor -1) { 
         let sub_tasks = keys[start..].iter().map(|key| tokio::task::spawn(async move { .. tokio_magic_here}));
         allHandles.push(sub_tasks);
         break; //since we are in the end loop here 
   }
   //otherwise continue like this
   let sub_tasks = keys[start..end].iter().map(|key| tokio::task::spawn(async move { .. tokio_magic_here}));
   allHandles.push(sub_tasks);
   start  = end; // we have to start from the point where we left off previously ! 
}
...
...

Reply:
there will be not only context switching overhead in case we need to transfer data between os threads, but also the overhead for managing async task itself as the scheduler will have to check more tasks in that case and while that is not very expensive with the high number it could affect perf.
Also, while you're right that splitting many cpu tasks should perform better when running in parallel than on single thread you should consider next points:
- it'll only be beneficial in case the operation itself is in some degree longer than the costs of passing data to another thread otherwise it'll actually worse the perf
- even if it is the case you should also consider how the whole application is working. We want to measure the overall performance of the app not only part of it. But when you are loading the os threads with only batching logic it could prevent that threads from doing any other staff i.e. handling other requests that could affect result numbers
- real word behaviour could significantly differ from what we think is going on so the actual tests numbers is the best judge. We need to just figure out that we're testing that right

Reply:
I strong agree to the above points, which must be considered for a educated decision here, since we are targeting optimal scale It won't be very efficient to have our values passed from one thread to another and save those CPU cycles for actually computing, rather than loosing them shifting values from one thread to another.

I think we will get better picture once we test for large number of batched queries in a schema file (and benchmarking the crucial sections to figure out areas to target for modification), since currently we are just testing out on single DataLoader here.

After running these tests it will make things clear for us, so for now it would be better that maybe we don't rely on the first approach absolutely, adding to approach I studied the primer material on runtime here, it can make theoretical sense to have two seprate runtime instances, first one obviously for handling the server query, another one responsible for solely managing the CPU intensive tasks such as one present in Loader section, let me know what you think about it ?